p2p/server.py

```diff
@@ -0,0 +1,156 @@
+        try:
+            await self._handle_peer(peer)
+        finally:
+            self._running_peers.remove(peer)
+
+    async def _handle_peer(self, peer: ETHPeer) -> None:
+        while True:
+            try:
+                cmd, msg = await peer.read_sub_proto_msg(self.cancel_token)
+            except OperationCancelled:
+                break
+
+    async def stop(self) -> None:
+        self.logger.info("Closing server...")
+        self.cancel_token.trigger()
+        self.peer_pool.unsubscribe(self)
+        while self._running_peers:
+            self.logger.debug("Waiting for %d running peers to finish", len(self._running_peers))
+
+    async def run(self) -> None:
+        self.logger.info("Running server...")
+        loop = asyncio.get_event_loop()
+        factory = asyncio.start_server(self.receive_handshake, *self._server_address)
+        asyncio.ensure_future(factory)
+        
+        while not self.cancel_token.triggered:
+            try:
+                loop.run_forever()
+            except OperationCancelled:
+                break
+
+    async def receive_handshake(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter) -> None:
+        # Use reader to read the auth_init msg until EOF
+        msg = await reader.read()
+
+        # Use HandshakeResponder.decode_authentication(auth_init_message) on auth init msg
+        ephem_pubkey, initiator_nonce, initiator_pubkey = decode_authentication(msg, self.privkey)
+
+        # Get remote's address 
+        ip, udp, _, _ = writer.get_extra_info("peername")
+        remote_address = Address(ip, udp)
+
+        # Create a `HandshakeResponder(remote: kademlia.Node, privkey: datatypes.PrivateKey)` instance
+        initiator_remote = Node(initiator_pubkey, remote_address)
+        responder = HandshakeResponder(initiator_remote, self.privkey)
+
+        # Call `HandshakeResponder.create_auth_ack_message(nonce: bytes)` to create the reply
+        responder_nonce = secrets.token_bytes(HASH_LEN) 
+        auth_ack_msg = responder.create_auth_ack_message(nonce=responder_nonce) 
+        auth_ack_ciphertext = responder.encrypt_auth_ack_message(auth_ack_msg)
+
+        # Use the `writer` to send the reply to the remote
+        writer.write(auth_ack_ciphertext)
+        await writer.drain()
+        
+        # Use `HandshakeResponder.derive_shared_secrets()` and use the return values to instantiate a `Peer` instance
+        aes_secret, mac_secret, egress_mac, ingress_mac = responder.derive_secrets(
+            initiator_nonce=initiator_nonce,
+            responder_nonce=responder_nonce,
+            remote_ephemeral_pubkey=ephem_pubkey,
+            auth_init_ciphertext=msg,
+            auth_ack_ciphertext=auth_ack_ciphertext
+        )
+
+        # Create and register peer in peer_pool
+        peer = (aes_secret, mac_secret, initiator_nonce, ephem_pubkey)
+        eth_peer = ETHPeer(
+                remote=initiator_remote, privkey=None, reader=None,
```

Comment:
@gsalgado I'm confused as to how we instantiate this peer to pass to the peer_Pool. How do we know the peer's privkey?

Reply:
We don't.  It looks like we're supposed to use our own `privkey` here.  cc @gsalgado to confirm.

Reply:
ISTM that ETHPeer doesn't use the `privkey` anymore, so we should just remove it

Reply:
@gsalgado Looks that way to me, should I go ahead and remove the `privkey` arg from the BasePeer class?

Reply:
Col, I'll just use the available `privkey` here for now and then in a following pr remove the `privkey` arg from BasePeer class, if that sounds good?

Reply:
It is actually used in `P2PProtocol.send_handshake()`, so yeah, just pass the pool's privkey here